src/geom/util/GeometryFixer.cpp

```diff
@@ -0,0 +1,337 @@
+#include <geos/geom/Geometry.h>
+#include <geos/geom/GeometryFactory.h>
+#include <geos/operation/overlayng/OverlayNG.h>
+#include <geos/operation/overlayng/OverlayNGRobust.h>
+#include <geos/operation/buffer/BufferOp.h>
+#include <geos/operation/valid/RepeatedPointRemover.h>
+#include <geos/util/UnsupportedOperationException.h>
+
+using geos::operation::valid::RepeatedPointRemover;
+
+namespace geos {
+namespace geom { // geos.geom
+namespace util { // geos.geom.util
+
+/* public static */
+std::unique_ptr<Geometry>
+GeometryFixer::fix(const Geometry* p_geom)
+{
+    GeometryFixer fix(p_geom);
+    return fix.getResult();
+}
+
+/* public */
+void
+GeometryFixer::setKeepCollapsed(bool p_isKeepCollapsed)
+{
+    isKeepCollapsed = p_isKeepCollapsed;
+}
+
+/* public */
+std::unique_ptr<Geometry>
+GeometryFixer::getResult()
+{
+    /**
+     *  Truly empty geometries are simply copied.
+     *  Geometry collections with elements are evaluated on a per-element basis.
+     */
+    if (geom->getNumGeometries() == 0) {
+        return geom->clone();
+    }
+
+    GeometryTypeId typeId = geom->getGeometryTypeId();
+
+    if (typeId == GEOS_POINT)              return fixPoint(static_cast<const Point*>(geom));
+    //  LinearRing must come before LineString
+    if (typeId == GEOS_LINEARRING)         return fixLinearRing(static_cast<const LinearRing*>(geom));
+    if (typeId == GEOS_LINESTRING)         return fixLineString(static_cast<const LineString*>(geom));
+    if (typeId == GEOS_POLYGON)            return fixPolygon(static_cast<const Polygon*>(geom));
+    if (typeId == GEOS_MULTIPOINT)         return fixMultiPoint(static_cast<const MultiPoint*>(geom));
+    if (typeId == GEOS_MULTILINESTRING)    return fixMultiLineString(static_cast<const MultiLineString*>(geom));
+    if (typeId == GEOS_MULTIPOLYGON)       return fixMultiPolygon(static_cast<const MultiPolygon*>(geom));
+    if (typeId == GEOS_GEOMETRYCOLLECTION) return fixCollection(static_cast<const GeometryCollection*>(geom));
+    throw geos::util::UnsupportedOperationException("GeometryFixer::getResult called on unknown geometry type");
+}
+
+/* private */
+std::unique_ptr<Point>
+GeometryFixer::fixPoint(const Point* p_geom)
+{
+    std::unique_ptr<Point> pt = fixPointElement(p_geom);
+    if (pt == nullptr)
+        return factory->createPoint();
+    else
+        return pt;
+}
+
+/* private */
+std::unique_ptr<Point>
+GeometryFixer::fixPointElement(const Point* p_geom)
+{
+    if (p_geom->isEmpty() || ! isValidPoint(p_geom)) {
+        return std::unique_ptr<Point>(nullptr);
+    }
+    return p_geom->clone();
+}
+
+/* private static */
+bool
+GeometryFixer::isValidPoint(const Point* pt)
+{
+    const Coordinate* p = pt->getCoordinate();
+    return p->isValid();
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixMultiPoint(const MultiPoint* p_geom)
+{
+    std::vector<std::unique_ptr<Point>> pts;
+    for (std::size_t i = 0; i < p_geom->getNumGeometries(); i++) {
```

Comment:
Unrelated to that PR, but it looks to me that the GEOS C++ geometry classes could benefit from being used with C++11 range-based loops. I did that in OGR C++ geom classes and that might iterating over child geometries of collections much cleaner. You do: ```for( const Point* pt: *p_geom ) { ... }```

Reply:
Yes, every time I do one of these I go "oooh, do we have an iterator on the collection type..." and then I go "oh, hm, we don't... don't get distracted, finish what you started". I'm not sure if there's a reason we don't (@dbaston ?) but it would certain be prettier frequently.

Reply:
I thought we had one on `GeometryCollection` but not typed variants for the subclasses.